### PR TITLE
gscan2pdf: new port + its perl dependencies

### PR DESCRIPTION
**Pull request by tenzap**
I'm not sure if it is accepted to commit all the dependencies at the same time. But since these dependencies are nested I thought it would be better that way.

gscan2pdf: A GUI to produce PDFs or DjVus from scanned documents

Perl dependencies: new ports:
- p5-carp-always
- p5-filesys-df
- p5-goocanvas2
- p5-graphics-tiff
- p5-gtk3-imageview
- p5-gtk3-simplelist
- p5-image-png-libpng
- p5-image-sane
- p5-pdf-builder
- p5-set-intspan

###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 10.13.6 17G65
Xcode 9.4.1

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by reneeotten** [comment]
> I'm not sure if it is accepted to commit all the dependencies at the same time. But since these dependencies are nested I thought it would be better that way.

it's fine to have them in the same PR, but each port should be added as a separate commit.

**Comment by tenzap** [comment]
> > I'm not sure if it is accepted to commit all the dependencies at the same time. But since these dependencies are nested I thought it would be better that way.
> 
> it's fine to have them in the same PR, but each port should be added as a separate commit.

Ok, it's split.

**Comment by tenzap** [comment]
@reneeotten , any news or comment on this?

File: graphics/gscan2pdf/Portfile
Comment (reneeotten) [inline graphics/gscan2pdf/Portfile:4]: no need for newline here

File: graphics/gscan2pdf/Portfile
Comment (reneeotten) [inline graphics/gscan2pdf/Portfile:62]: are these optional dependencies used / linked to in case the user has them installed (assuming they're available in MacPorts)? If so, that's not what we want to get reproducible builds. In that case you'll either want to make sure the dependencies are always installed or never used even if they are present.

File: perl/p5-carp-always/Portfile
Comment (reneeotten) [inline perl/p5-carp-always/Portfile:6]: this should now include also 5.34 (applies to other ports as well)

File: graphics/gscan2pdf/Portfile
Comment (reneeotten) [inline graphics/gscan2pdf/Portfile:12]: this is the default since MacPorts v2.7 so the line can be removed (also in the other ports)

File: graphics/gscan2pdf/Portfile
Comment (reneeotten) [inline graphics/gscan2pdf/Portfile:50]: no need for additional newline

**Review by reneeotten — CHANGES_REQUESTED** [review]
@tenzap I was hoping a more experience Perl user would review this PR but it doesn't look like that's going to happen... I've left a few more comments, most of them trivial but the one for the optional dependencies needs to be checked for sure. Once that's done I am happy to merge and we can resolve possible issues when they come up.

File: graphics/gscan2pdf/Portfile
Comment (tenzap) [reply graphics/gscan2pdf/Portfile:62]: According to gscan2pdf, they are optional. They are not mandatory to build, neither to work but some features of the SW will be missing.

- pdftk is used for pdf encryption/decryption. **Pulls the full java (for pdftk-java package)**
- djvulibre to save as djvu
- xdg-util Required for Email as PDF.
- sane-frontends is only needed for scanadf frontend.
- For unpaper it will warn if the program is missing, and functionality is greyed out. **Pulls ~20 dependencies**
- For tesseract/gocr/cuneiform it will detect those installed and propose to choose one of them (if one wants to use OCR) If no OCR is present, the functionality is greyed out. The problem is see here is that tesseract for example **is a very large dependency** and we don't necessarily want it. Moreover, there is also gocr and cuneiform which do the same job. So I'm not sure it should be a strong dependency for the software.

I don't really now how to handle this. Would that be ok?
- put `xdg-utils`, `djvulibre` & `sane-frontends` as mandatory dependency (they don't look that big)
- **and** add these variants? `+tesseract +gocr +cuneiform +pdtk +unpaper`

Comment (reneeotten) [reply graphics/gscan2pdf/Portfile:62]: it all depends; if the program doesn't link to any of the other packages but only uses them if available at runtime then it doesn't matter and you can just leave them out. If not, the you'll do something about it as I said above. So you can check whether the resulting binary links to anything (using `otool -L <filename>` when you have some of the packages installed.

Comment (tenzap) [reply graphics/gscan2pdf/Portfile:62]: BTW gscan2pdf is in perl, so no binary.

Comment (tenzap) [reply graphics/gscan2pdf/Portfile]: I added a note.

File: graphics/gscan2pdf/Portfile
Comment (reneeotten) [reply graphics/gscan2pdf/Portfile]: please remove that line in every Portfile in this PR

Comment (tenzap) [reply graphics/gscan2pdf/Portfile]: done, haven't seen them

**Comment by reneeotten** [comment]
did this build for you locally? because if I try I get this:
```
Error: Failed to activate gscan2pdf: Image error: /opt/local/lib/perl5/5.28/darwin-thread-multi-2level/perllocal.pod is being used by the active ossp-uuid port.  Please deactivate this port first, or use 'port -f activate gscan2pdf' to force the activation.
DEBUG: Error code: registry::image-error
DEBUG: Backtrace: Image error: /opt/local/lib/perl5/5.28/darwin-thread-multi-2level/perllocal.pod is being used by the active ossp-uuid port.  Please deactivate this port first, or use 'port -f activate gscan2pdf' to force the activation.
    invoked from within
"throw registry::image-error "Image error: $file is being used by the active [$owner name] port.  Please deactivate this port first, or use 'port -f ac..."
    ("foreach" body line 45)
    invoked from within
"foreach file $imagefiles {
                set srcfile "${extracted_dir}${file}"

                # To be able to install links, we test if we can lst..."
    invoked from within
"registry::write {
            foreach file $imagefiles {
                set srcfile "${extracted_dir}${file}"

                # To be able to instal..."
    invoked from within
"try {
        registry::write {
            foreach file $imagefiles {
                set srcfile "${extracted_dir}${file}"

                # To be ..."
    (procedure "_activate_contents" line 21)
    invoked from within
"_activate_contents $requested $rename_list"
    (procedure "portimage::activate" line 63)
    invoked from within
"registry_activate $subport $_inregistry_version $_inregistry_revision $_inregistry_variants $optionlist"
    (procedure "portactivate::activate_main" line 20)
    invoked from within
"$procedure $targetname"
Error: See /opt/local/var/macports/logs/_Users_renee_Software_macports-ports_graphics_gscan2pdf/gscan2pdf/main.log for details.
Error: Follow https://guide.macports.org/#project.tickets if you believe there is a bug.
Error: Processing of port gscan2pdf failed
```

**Comment by tenzap** [comment]
Yes I did. On 10.13. It worked. Didn't try with today's fixes which were only cosmetic

**Comment by reneeotten** [comment]
> Yes I did. On 10.13. It worked. Didn't try with today's fixes which were only cosmetic

Please check if you have the "ossp-uuid" port installed and if not please do. I'm pretty sure the conflict is real and this should be resolved before merging.

**Comment by tenzap** [comment]
I can reproduce now. However I don't understand how this should be fixed.
Should I deactivate the ossp-uuid port and reactivate it after installation? Should I add a conflict in the portfile? Thanks

**Comment by tenzap** [comment]
Seems the perllocal.pod should not be installed, neither by gscan2pdf, nor by ossd-uuid.
Actually wouldn't the correct fix, if this file should never be installed by a port, be to have that in the perl portgroup ?

File: graphics/gscan2pdf/Portfile
Comment (reneeotten) [inline graphics/gscan2pdf/Portfile:54]: there is no reason to do a `glob` here since there is only one file that you want to remove; also as `port lint` points out you shouldn't use `eval`. So just remove the one file, presumably with (but please double-check as I haven't tried it):
``` file delete ${destroot}${prefix}/lib/perl5/${perl5.major}/darwin-thread-multi-2level/perllocal.pod```

and please add a comment of why this was needed (i.e., conflict with the a file that has the same name but belongs to the `ossp-uuid` port) and perhaps a link to the Trac tickets you mentioned.

**Comment by reneeotten** [comment]
> Seems the perllocal.pod should not be installed, neither by gscan2pdf, nor by ossp-uuid. Actually wouldn't the correct fix, if this file should never be installed by a port, be to have that in the perl portgroup (See https://trac.macports.org/ticket/39384#comment:5) ?

I have no idea - I am not a Perl user and haven't looked into the details; you could comment in the ticket you mention and see if there are any experienced Perl users and/or package maintainers around who do know.

File: graphics/gscan2pdf/Portfile
Comment (tenzap) [reply graphics/gscan2pdf/Portfile:54]: Well I copied the code from other ports that do it that way. I can't check in how many ports this is used right now but I got the feeling that this is standard in some perl ports. An advice of perl port maintainers would be welcome on what to do.

Comment (reneeotten) [reply graphics/gscan2pdf/Portfile:54]: I  just `grep`-ed for `poerllocal.pod` in the MacPorts tree and there are no ports that use `eval` as far as I can tell, so you'll need to correct that for sure. And since here there is only one file that you need to remove I would just specify it explicitly

Comment (tenzap) [reply graphics/gscan2pdf/Portfile:54]: Ok. Thanks for looking.
I created ticket https://trac.macports.org/ticket/64444 in the meantime.

Comment (tenzap) [reply graphics/gscan2pdf/Portfile]: Updated as per your recommendation (it installs fine)

**Comment by reneeotten** [comment]
@tenzap any idea why the CI fails here? That should likely be resolved before this PR can be merged or the same will happen on the buildbots.

I see that you didn't check off ` tried a full install with sudo port -vst install?` in the PR checklist. Please make sure to do so to double-check it actually builds on a more-or-less clean system and check the CI logs to see what the issue is.

**Comment by tenzap** [comment]
I don't know why it constantly fails in CI. By looking at the logs is seems it is because the dependencies of gscan2pdf fail (p5*-image-sane dependency of gscan2pdf is not there at the time gscan2pdf is built?). If the builder tries to build gscan2pdf before all the perl modules, it will fail, indeed. The order of the commits is important.

Are you able to build it?

I ran `port -vst install gscan2pdf` without problem.

**Comment by reneeotten** [comment]
> Are you able to build it? I ran `port -vst install gscan2pdf` without problem.

me too so it seems fine. I'll merge this PR and if anything comes up we can look into it - thanks for the submission and your patience!

**Review by reneeotten — APPROVED** [review]

**Pull request merged**
